### PR TITLE
SIL: Lower captures to boxes with an appropriate generic context.

### PR DESCRIPTION
**Pull request by jckarter**
Officially kick SILBoxType over to be "nominal" in its layout, with generic layouts structurally parameterized only by formal types. Change SIL to lower a capture to a nongeneric box when possible, or a box capturing the enclosing generic context when necessary.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test and merge

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/SIL/SILType.cpp [hunk 1]
```diff
@@ -555,3 +555,29 @@ SILType SILType::getReferentType(SILModule &M) const {
...
+  auto fieldTy = getLayout()->getFields()[index].getLoweredType();
+  
+  // Apply generic arguments if the layout is generic.
+  if (!getGenericArgs().empty()) {
+    // FIXME: Map the field type into the layout's generic context because
+    // SIL TypeLowering currently expects to lower abstract generic parameters
+    // with a generic context pushed, but nested generic contexts are not
+    // supported by TypeLowering. If TypeLowering were properly
+    // de-contextualized and plumbed through the generic signature, this could
+    // be avoided.
```
Comment (slavapestov) [inline lib/SIL/SILType.cpp]: If you want you can take a look at this old PR and see if you can salvage anything: https://github.com/apple/swift/pull/5524/commits

File: lib/Sema/TypeCheckGeneric.cpp [hunk 2]
```diff
@@ -170,7 +170,7 @@ Type CompleteGenericTypeResolver::resolveDependentMemberType(
   // Resolve the base to a potential archetype.
   auto basePA = Builder.resolveArchetype(baseTy);
   assert(basePA && "Missing potential archetype for base");
-  basePA = basePA->getRepresentative();
+  basePA = basePA->getArchetypeAnchor();
```
Comment (slavapestov) [inline lib/Sema/TypeCheckGeneric.cpp]: Why this change?

File: lib/IRGen/GenType.cpp [hunk 3]
```diff
@@ -1125,14 +1125,19 @@ CanType TypeConverter::getExemplarType(CanType contextTy) {
   // FIXME: A generic SILFunctionType should not contain any nondependent
```
Comment (slavapestov) [inline lib/IRGen/GenType.cpp]: In fact a generic silfunctiontype should not contain archetypes at all. I suspect this entire block of code is dead

**Comment by jckarter** [comment]
@swift-ci Please smoke test Linux

File: lib/Sema/TypeCheckGeneric.cpp (hunk 2 above)
Comment (jckarter) [inline lib/Sema/TypeCheckGeneric.cpp]: Doug had suggested it while I was trying to work through some same type issues. It doesn't appear to be necessary, so I can remove it.

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test
